### PR TITLE
CSV download permission download

### PR DESCRIPTION
**Pull request by xnetsrak**
## References

* #0000

### CODE REVIEW

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.Application/Handlers/Permissions/GetPermissionRelationsHandler.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    public async Task<Stream> Handle(GetPermissionRelationsCommand request, CancellationToken cancellationToken)
+    {
+        var records = await _permissionRelationService.BuildRelationRecordsAsync().ConfigureAwait(false);
+
+        return WriteRecordsToStream(records
+            .OrderBy(e => e.MarketRole)
+            .ThenBy(e => e.Permission));
+    }
+
+    private static Stream WriteRecordsToStream(IEnumerable<PermissionRelationRecord> records)
```
Comment (michaelmajgaard) [inline source/marketparticipant/Energinet.DataHub.MarketParticipant.Application/Handlers/Permissions/GetPermissionRelationsHandler.cs:46]: I'd use StreamWriter here instead I think:
```c#
var stream = new MemoryStream();
using var streamWriter = new StreamWriter(stream, Encoding.UTF8, leaveOpen: true);
streamWriter.WriteLine("PermissionName;MarketRole;UserRole");

foreach (var record in records)
{
    streamWriter.WriteLine($"{record.Permission};{record.MarketRole};{record.UserRole}");
}

streamWriter.Flush();
return stream;
```

Comment (xnetsrak) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.Application/Handlers/Permissions/GetPermissionRelationsHandler.cs:46]: Why ?

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.Application/Services/PermissionRelationService.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+        var allMarketRoles = Enum.GetNames<EicFunction>();
+
+        var records = new List<PermissionRelationRecord>();
+
+        foreach (var permission in allPermissions)
+        {
+            var userRoles = allUserRoles.Where(x => x.Permissions.Contains(permission.Id)).ToList();
+
+            if (userRoles.Any())
```
Comment (michaelmajgaard) [inline source/marketparticipant/Energinet.DataHub.MarketParticipant.Application/Services/PermissionRelationService.cs:52]: You can skip this check

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.EntryPoint.WebApi/Controllers/PermissionController.cs [hunk 3]
```diff
@@ -88,5 +89,26 @@ public async Task<ActionResult<IEnumerable<AuditLogDto<PermissionAuditedChange>>
...
+
+        [HttpGet("relation")]
+        [AuthorizeUser(PermissionId.UserRolesManage)]
+        [Produces("application/octet-stream")]
+        [ProducesResponseType(typeof(FileResult), StatusCodes.Status200OK)]
+        public async Task<ActionResult> GetPermissionsRelationAsync()
+        {
+            if (!_userContext.CurrentUser.IsFas)
+            {
+                Response.StatusCode = StatusCodes.Status401Unauthorized;
```
Comment (michaelmajgaard) [inline source/marketparticipant/Energinet.DataHub.MarketParticipant.EntryPoint.WebApi/Controllers/PermissionController.cs:101]: This is redundant when using `return Unauthorized();`

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.Tests/Services/PermissionRelationServiceTest.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+        var permissionRelationService = new PermissionRelationService(mockPermissionRepository.Object, mockUserRoleRepository.Object);
+
+        // Act
+        var records = (await permissionRelationService.BuildRelationRecordsAsync()).ToList();
+
+        // Assert
+        var testPermissions = BuildTestPermissions().ToList();
+        var marketRoles = Enum.GetValues<EicFunction>();
+        Assert.NotNull(records);
```
Comment (michaelmajgaard) [inline source/marketparticipant/Energinet.DataHub.MarketParticipant.Tests/Services/PermissionRelationServiceTest.cs:76]: Doesn't look like the return type is nullable

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.Tests/Services/PermissionRelationServiceTest.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        Assert.NotEmpty(records);
```
Comment (michaelmajgaard) [inline source/marketparticipant/Energinet.DataHub.MarketParticipant.Tests/Services/PermissionRelationServiceTest.cs:77]: This is also a bit redundant as you assert the count is 26 on the next line

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.EntryPoint.WebApi/Controllers/PermissionController.cs (hunk 3 above)
Comment (xnetsrak) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.EntryPoint.WebApi/Controllers/PermissionController.cs:101]: Copied from CSV download in ImbalancePrices Repo :D

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.Application/Handlers/Permissions/GetPermissionRelationsHandler.cs (hunk 1 above)
Comment (michaelmajgaard) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.Application/Handlers/Permissions/GetPermissionRelationsHandler.cs:46]: Looks like you want the result returned as a stream.

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.Application/Services/PermissionRelationService.cs (hunk 2 above)
Comment (michaelmajgaard) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.Application/Services/PermissionRelationService.cs:52]: No reason to guard the `foreach` behind a `.Any()` is what I mean.

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.EntryPoint.WebApi/Controllers/PermissionController.cs (hunk 3 above)
Comment (michaelmajgaard) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.EntryPoint.WebApi/Controllers/PermissionController.cs]: Then fix it there too :D

File: source/marketparticipant/Energinet.DataHub.MarketParticipant.Application/Services/PermissionRelationService.cs (hunk 2 above)
Comment (michaelmajgaard) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.Application/Services/PermissionRelationService.cs:52]: Ah nvm, I'm stupid. I see your else case now XD

Comment (xnetsrak) [reply source/marketparticipant/Energinet.DataHub.MarketParticipant.Application/Services/PermissionRelationService.cs:52]: You understand programming ?

**Review by michaelmajgaard — APPROVED** [review]

**Pull request merged**